docs/dev/design/secure-api-access-from-builders.rst

```diff
@@ -0,0 +1,157 @@
+
+Non-goals
+---------
+
+- Migrate builds to use API V3
+- Implement this mechanism in API V3
+- Expose it to users
+
+All these changes can be made in the future, if needed.
+
+Current state
+-------------
+
+Currently, we access the API V2 from the builders using the credentials of the "builder" user.
+This user is a superuser, it has access to all projects,
+write access to the API, access to restricted endpoints, and restricted fields.
+
+The credentials are hardcoded in our settings file,
+so if there is a vulnerability that allows users to have access to the settings file,
+the attacker will have access to the credentials of the "builder" user,
+giving them full access to the API and all projects.
+
+Proposed solution
+-----------------
+
+Instead of using the credential of a super user to access the API,
+we will create a temporal token attached to a project, and one of the owners of the project.
+This way this token will have access to the given project only for a limited period of time.
+
+This token will be generated from the webs,
+and passed to the builders via the celery task,
+where it can be used to access the API.
+Once the build has finished, this token will be revoked.
+
+Technical implementation
+------------------------
+
+We will use the rest-knox_ package,
+this package is recommended by the DRF documentation,
+since the default token implementation of DRF is very basic,
+some relevant features of knox are:
+
+- Support for several tokens per user.
+- Tokens are stored in a hashed format in the database.
+  We don't have access the tokens after they are created.
+- Tokens can hava an expiration date.
+- Tokens can be created with a prefix (rtd_xxx) (unreleased)
+- Support for custom token model (unreleased)
+
+We won't expose the token creation view directly,
+since we can create the tokens from the webs,
+and this isn't exposed to users.
+
+The view to revoke the token will be exposed,
+since we need it to revoke the token once the build has finished.
+
+From the API, we just need to add the proper permission and authentication classes
+to the views we want to support.
+
+To differentiate from a normal user and a token authed user,
+we will have access to the token via the ``request.auth`` attribute in the API views,
+this will also be used to get the attached projects to filter the querysets.
+
+The knox package allows us to provide our own token model,
+this will be useful to add our own fields to the token model.
+Fields like the projects attached to the token,
+or access to all projects the user has access to, etc.
+
+.. _rest-knox: https://james1345.github.io/django-rest-knox/
+
+Flow
+----
+
+The flow of creation and usage of the token will be:
+
+- Create a token from the webs when a build is triggered.
+  The triggered project will be attached to the token,
+  if the build was triggered by a user, that user will be attached to the token,
+  otherwise the token will be attached to one of the owners of the project.
```

Comment:
I don't love this requirement, but seems fine... Does attaching it to a project user make more sense than just using a dummy user of some kind?

Reply:
Aside from _needing_ some way to attach a token scope to a project rather than a user, do we have other cases where we would need to customize the functionality of `rest-knox`?

Token expiry is covered, what about scopes? Do we need them only per-user? Or also per-team or per-organization?

Do we need to limit rest-knox tokens beyond what they are originally able to do? Can they do _more_ than what a build needs?

Just asking these questions to understand how well the package fits, and if it fits, that's great :+1: If it needs some level of customization or we have to accept significant gaps, then there can be a motivation to implement our own token authentication.

I'm only knowledgeable of "rolling your own token authentication" from django-ninja, but in this case, it was easy because django-ninja was already designed for it.

Reply:
We can attach an organization to the token if needed, but knox requires you to always have a user attached to it.